### PR TITLE
Query: Introduce custom query root for TVF

### LINKED ISSUE #20146: Query: Introduce an Expression for query root
Currently we use ConstantExpression of EntityQueryable as query root.
EntityQueryable is an implementation of IQueryable which wraps current expression trees and allows it to be enumerated using IQueryProvider. 
- IQueryProvider is not needed during expression tree processing in anyway. It is a bootstrapper to execute query.
- By introducing query root as expression we can streamline our implementation since we deal with ExpressionVisitors everywhere. (We pattern match constant of entity queryable everywhere
- Allows derived query roots which can be visited by individual visitors also.

Split off from #18923

### PR DESCRIPTION
**Pull request by smitpatel**
- Avoids having to do special processing in other visitors
- Visit's method's argument correctly in other visitor (this means that if there is a navigation expansion required for a parameter it will be done automatically)
- Remove Nav expansion factory not needed anymore
- Rename ExpressionPrinter.VisitList to VisitCollection and take IReadOnlyCollection, that is what most expression uses.
- Remove unnecessary exception message. They are part of a larger work item.

Resolves #20146

### CODE REVIEW

**Comment by roji** [comment]
@smitpatel confirmed that with your branch all my tests are green, thanks!

File: test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs [hunk 1]
```diff
@@ -1410,31 +1410,32 @@ select new
...
         {
             using (var context = CreateContext())
             {
                 var query = from c in context.Customers
                             select new { c.Id, products = context.GetTopSellingProductsForCustomer(c.Id).ToList() };
 
-                Assert.Contains(
-                    RelationalStrings.DbFunctionProjectedCollectionMustHavePK("GetTopSellingProductsForCustomer"),
-                    Assert.Throws<InvalidOperationException>(() => query.ToList()).Message);
+                //Assert.Contains(
```
Comment (maumar) [inline test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs]: remove

File: test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs [hunk 2]
```diff
@@ -1410,31 +1410,32 @@ select new
...
             using (var context = CreateContext())
             {
                 var query = (from c in context.Customers
                              select new { c.Id, orders = context.GetCustomerOrderCountByYear(c.Id) });
 
-                Assert.Contains(
-                    RelationalStrings.DbFunctionCantProjectIQueryable(),
-                    Assert.Throws<InvalidOperationException>(() => query.ToList()).Message);
+
+                //Assert.Contains(
```
Comment (maumar) [inline test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs]: same here

File: test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs (hunk 1 above)
Comment (smitpatel) [reply test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs]: The linked issues will add exception message and these tests will verify exception only hence I have not removed them.

**Pull request merged**